exampleCourse/questions/examplesRGetData/server.py

```diff
@@ -0,0 +1,55 @@
+# Enable R computation
+import rpy2.robjects as robjects
+
+def generate(data):
+    
+    # Get list of data sets in base R  
+   
+    values = robjects.r('''
+    # Get list of data sets in base R  
+    datasets = c('CO2', 'attitude', 'USJudgeRatings', 'pressure', 'warpbreaks', 'esoph', 'ToothGrowth')
+    
+    # Choose a data set
+    ds_name = sample(datasets, size = 1)
+
+    # Retrieve the data set
+    df_picked = get(ds_name)
+
+    # Obtain the dimensions
+    df_dim = dim(df_picked)
+
+    ds_nrow = df_dim[1]
+    ds_ncol = df_dim[2]
+
+    # Cap observation number
+    if(ds_nrow > 15) {
+       ds_nrow = 15
+    }
+
+    df = knitr::kable(head(df_picked, ds_nrow),
+                 align = rep('c', ds_ncol), format = 'html', row.names = FALSE)
+
+    list(params = list(display_data = df, 
+                       bad_row_low = ds_nrow + sample(-8:-1, 1),
+                       bad_row_medium = ds_nrow + sample(1:3, 1),
+                       bad_row_high = ds_nrow + sample(4:6, 1),
+                       bad_row_highest = ds_nrow + sample(6:9, 1),
+                       bad_col_low = ds_ncol + sample(-2:-1, 1), 
+                       bad_col_medium = ds_ncol + sample(1:3, 1),
+                       bad_col_high = ds_ncol + sample(4:7, 1), 
+                       bad_col_highest = ds_ncol + sample(7:10, 1)),
+         ans    = list(ds_nrow = ds_nrow, ds_ncol = ds_ncol))
```

Comment:
The above randomization strategy means that the true answer is always the second-lowest answer out of the five options, if I'm understanding correctly.

An alternative secure version (written in Python) is:
```
# calculate the true answer somehow
true_answer = 42

# parameters for generating distractors
number_answers = 5  # 1 true answer plus 4 distractors
min_gap = 1   # minimum distance between answers
max_gap = 4   # maximum distance between answers

# generate distinct random answer offsets from random gaps between answers
# this is not essential to the provably-secure generation (e.g., could use accept/reject instead)
gaps = np.random.randint(min_gap, max_gap, number_answers)
offsets = np.cumsum(gaps)

# choose one of the answers to be the true answer
true_answer_index = np.random.randint(number_answers)

# offset of the true answer from the smallest answer
true_answer_offset = offsets[true_answer_index]

# calculate the answers by computing the offsets from the known true answer
answers = true_answer + offsets - true_answer_offset

# eliminate the true answer from the answers array to get just the distractors
distractors = np.delete(answers, true_answer_index)

# the list of distractors are now in distractors[0], distractors[1], etc.
```

I believe that the above algorithm is provably secure, in the sense that:
```
Prob(c_i = T | v_1,...,v_n) = Prob(c_i = T) = 1/n for all i = 1,...,n
```
where
```
n = number of available answers
v_1,...,v_n = numerical values of each answer
c_1,...,c_n = T/F for each answer, indicating whether it's a correct answer
```
That is, the probability that a particular answer is correct is independent of the numerical value for that answer, and it is also independent of the numerical values of the other answers.

Proving this seems like a straightforward application of Bayes, assuming an uniformative prior given by a uniform likelihood over the reals (technically, an improper prior) that satisfies translation invariance. That is, translation invariance of the prior means that
```
P(v_1,...,v_n | c_i = T) / P(v_1,...,v_n) = 1
```
and from the uniform selection of the true answer offset we have
```
P(c_i = T) = 1/n
```
so combining these with Bayes' theorem gives
```
P(c_i = T | v_1,...,v_n) = P(v_1,...,v_n | c_i = T) P(c_i = T) / P(v_1,...,v_n)
                         = (P(v_1,...,v_n | c_i = T) / P(v_1,...,v_n)) P(c_i = T)
                         = 1/n
```
which proves the result. We have a slight technical issue in that the translation invariance requires an improper prior (does not satisfy the law of total probability). This can be fixed using a uniform likelihood function `L(c_i = T, {v_1,...,v_n}) = 1` and Bayes' theorem written in likelihood form:
```
P(c_i = T | v_1,...,v_n) \propto L(c_i = T, {v_1,...,v_n}) P(c_i = T)
                         = 1/n
```
and then recognizing that this is already normalized gives `P(c_i = T | v_1,...,v_n) = 1/n`.

We can also write an equivalent algorithm for continuous-valued answers:
```
# calculate the true answer somehow
true_answer = 3.1415

# parameters for generating distractors
number_answers = 5  # 1 true answer plus 4 distractors
min_gap = 0.7   # minimum distance between answers
max_gap = 1.8   # maximum distance between answers

# generate distinct random answer offsets from random gaps between answers
# this is not essential to the provably-secure generation (e.g., could use accept/reject instead)
gaps = np.random.uniform(min_gap, max_gap, number_answers)
offsets = np.cumsum(gaps)

# choose one of the answers to be the true answer
true_answer_index = np.random.randint(number_answers)

# offset of the true answer from the smallest answer
true_answer_offset = offsets[true_answer_index]

# calculate the answers by computing the offsets from the known true answer
answers = true_answer + offsets - true_answer_offset

# eliminate the true answer from the answers array to get just the distractors
distractors = np.delete(answers, true_answer_index)

# the list of distractors are now in distractors[0], distractors[1], etc.
```

This satisfies the same provably-secure result as the discrete version above, using an identical proof.

I wonder whether we should include these algorithms in the `prairielearn.py` library so questions can easily use them? I feel like these methods must be in the literature somewhere, but I don't know where to look. If they aren't, then we should publish them 😄

Reply:
Surprisingly, there is little literature under ["question distractors numeric"](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C14&q=question+distractors+numeric&btnG=) and ["question distractors"](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C14&q=question+distractors&btnG=) on a precursory gaze. The majority of literature is focused on generating stems or re-using words within a given domain [c.f. "Automatic Generation of Cloze Question Distractors"](https://www.isca-speech.org/archive/L2WS_2010/papers/lw10_P2-11.pdf). I think this was largely left off of the MC formulation because the input is traditional `numeric`. 

Regarding the provably secure proof, let me think about it a bit more and get back to you.

Reply:
Go for it!

Reply:
One issue I think that creeps up with the distractor generation is the inability to set a lower bound for values to take on. As an example, the number of variables can now be negative.
![bounds](https://user-images.githubusercontent.com/833642/69452909-7a4de680-0d28-11ea-9d3a-4b03e212b29b.png)